### DISCUSSION TITLE
How to define a `Nullable` `Enum` property?

### DISCUSSION THREAD

**Discussion by KBroichhausen** [TypedProject/ts-express-decorators#2730] [Q&A] [answered]
Hey everyone,
given something like
```typescript
export enum BarEnum {
	Bar1 = "bar1",
	Bar2 = "bar2"
}

export class Foo {
	@Required(true, null)
	@Enum(BarEnum)
	public bar: BarEnum | null;
}
```
I get the following part in the `swagger,json`
```json
  "Foo": { "type": "object", "properties": { "bar": { "type": "string", "enum": ["bar1", "bar2"], "nullable": true, "minLength": 1 } }, "required": ["bar"] }
```

but if I actually use this model with a `null` value for `bar` as a post parameter in a controller I get a `PARAM_VALIDATION_ERROR` for this value with `[...] must be equal to one of the allowed values. Given value: null\n'`.

I don't really see the reason why. A workaround is to use something like
```typescript
export class Foo {
	@Required(true, null)
	@Enum(BarEnum, null)
	public bar: BarEnum | null;
}
```
but that adds `null` as value in the `enum` section of the schema which is bad because I use a code generator and this adds `"null"` (so basically null as a string) as a possible value for the enum.

**Comment by NachtRitter** [comment]
You can try to use `Any` decorator from `@tsed/schema` package like this

```ts
export enum BarEnum {
	Bar1 = "bar1",
	Bar2 = "bar2"
}

export class Foo {
	@Enum(BarEnum)
        @Any('string', 'null')
	public bar: BarEnum | null;
}
```

**Reply by KBroichhausen** [reply]
I tried it without success. Still get the same error `PARAM_VALIDATION_ERROR`.

**Reply by Romakita** [reply]
I suggest you to create repository code example please ;)

**Reply by KBroichhausen** [reply]
Ok, quickly created a default project with the cli, slightly modified and removed some unecessary stuff. https://github.com/KBroichhausen/nullable-enum
Should just be `npm install` and `npm run test`. Relevant code is in `HelloWorldController.ts` and the test file for it.
Test is failing because of `null` but works as soon as you activate the `// @Enum(BarEnum, null)` line (but I don't want to have `null` being part of the enum itself)

**Reply by Romakita** [reply, marked as answer]
Hello @KBroichhausen 
I tried that:

```ts
	@Required()
	@Allow(null)
	@Enum(BarEnum)
```

The json-schema is correctly generated, but the problem isn't related to Ts.ED. 

```json 
{
  "type": "object",
  "properties": {
    "bar": {
      "type": ["null", "string"],
      "enum": ["bar1", "bar2"], // and ["bar1", "bar2", null]
    }
  },
  "required": ["bar"]
}
```

AJV doesn't seems to support this kind of rules (https://github.com/ajv-validator/ajv/issues/824). I don't be able to fix that until ajv doesn't support correctly null value for an enum value.

I suggest you to add a default enum value like this:

```ts
export enum BarEnum {
	Bar1 = "bar1",
	Bar2 = "bar2",
	UNKNOWN = "UNKNOWN"
}

export class Foo {
	@Enum(BarEnum)
        @Default(BarEnum.UNKNOWN)
	public bar: BarEnum  = BarEnum.UNKNOWN;
}
```

See you
Romain

**Reply by KBroichhausen** [reply]
Oh, OK, Thank you very much. 
Most of the times if something wasn't working, I just had to dig deeper in the documentation. Didn't know this is really kind of an unsupported feature/bug.
To add a little bit more, partially unrequired, context: I use a code generator to create an angular library from the schema, then I use the enums to create options for a selector box (combined with a translation file). Eventually, it doesn't make much difference to use your suggestion with the default unknown value or use `@Enum(BarEnum, null)` which adds the null value to the schema and code generated enum => in both cases I had to filter that value out for the selector in the frontend.
As I have many more enums, I stick with the null value, so I don't have to deal with it on the backend. But I consider this more as a workaround, that's why I asked here for a proper solution.

**Reply by Romakita** [reply]
Yes I understand your point, it's consistent. I'm sorry to not have a real solution for the null value + enum. I just discovered this limitation and in fact, I never tried that have a null value on Enums before ;)

See you
